Add `output_dir` to cli argument

cyphersnake opened the pull request:
I want to use paperoni from qutebrowser directly in dropbox folder (and pocketbook), but for this I need to specify the folder in the arguments. I also refactored cli.rs a bit.

- Add `output_dir` to cli argument
    - This argument allows you to save output files in a special folder, not just current dir
- Refactor 'cli.rs'
    - Add `Builder` for `AppConfig`
    - Add `Error` instead separated panics
- Update dependencies

hipstermojo committed:
Update version

Commit:
Add `output_dir` to cli argument

- Add `output_dir` to cli argument
    - This argument allows you to save output files in a special folder, not just current dir
- Refactor 'cli.rs'
    - Add `Builder` for `AppConfig`
    - Add `Error` instead separated panics
- Upgrade dependencies

hipstermojo commented on src/cli.rs:
```diff
@@ -28,11 +71,21 @@ pub fn cli_init() -> AppConfig {
+            Arg::with_name("output_directory")
+                .long("output_directory")
+                .short("o")
+                .help("Directory for store output epub documents")
```
Change this to "Directory to store output epub documents"

hipstermojo commented on README.md:
```diff
@@ -48,15 +48,20 @@ USAGE:
+        --max_conn <max_conn>
+            The maximum number of concurrent HTTP connections when downloading articles. Default is 8
+
+    -o, --output_directory <output_directory>    Directory for store output epub documents
```
Update this after changing the help message

hipstermojo commented on src/cli.rs:
```diff
@@ -28,11 +71,21 @@ pub fn cli_init() -> AppConfig {
         )
+        .arg(
+            Arg::with_name("output_directory")
+                .long("output_directory")
```
The long flag should be in kebab case so change it to `output-directory`

hipstermojo commented on src/cli.rs:
```diff
@@ -60,143 +113,135 @@ pub fn cli_init() -> AppConfig {
+                match (
+                    arg_matches
+                        .values_of("urls")
+                        .and_then(|urls| urls.map(url_filter).collect::<Option<HashSet<_>>>()),
```
Use a `BTreeSet` instead of a `HashSet` to preserve order of the links

hipstermojo commented on src/cli.rs:
```diff
@@ -60,143 +113,135 @@ pub fn cli_init() -> AppConfig {
+                3 => LogLevel::Info,
+                4..=u64::MAX => LogLevel::Debug,
+            })
+            .is_logging_to_file(arg_matches.is_present("log-to_file"))
```
This should be "log-to-file"

hipstermojo commented on src/cli.rs:
```diff
@@ -28,11 +71,21 @@ pub fn cli_init() -> AppConfig {
+                .short("o")
+                .help("Directory for store output epub documents")
+                .conflicts_with("output_name")
+                .long_help("Directory for saving epub documents")
```
The `long_help` is not needed since the help message above is descriptive enough.

hipstermojo commented on src/cli.rs:
```diff
@@ -60,143 +113,135 @@ pub fn cli_init() -> AppConfig {
-    if arg_matches.is_present("verbosity") {
-        if !arg_matches.is_present("log-to-file") {
-            app_config.can_disable_progress_bar = true;
+    fn init_logger(self) -> Result<Self, Error> {
```
This function should still remain in the `log` crate but then return `Result<(),Error>`

cyphersnake replied:
If you don't mind, I will transfer the pure functionality to the crate, leaving this function here.

hipstermojo replied:
Okay. My mistake for not noticing that the builder pattern being used needs it to be this way.

hipstermojo commented on src/cli.rs:
```diff
@@ -60,143 +113,135 @@ pub fn cli_init() -> AppConfig {
+                        None
+                    }
+                };
+                match (
```
Make the values in the match expression their own variables.

hipstermojo commented on src/cli.rs:
```diff
@@ -1,13 +1,56 @@
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
```
Move this enum to the `error` module and name it `CliError`

hipstermojo commented on src/cli.rs:
```diff
@@ -1,13 +1,56 @@
+    #[error("No urls for parse")]
+    NoUrls,
+    #[error("No urls for parse")]
+    AppBuildError(#[from] AppConfigBuilderError),
```
This variant returns the same error message as `NoUrls` above. Should this be the case? I haven't used the `derive_builder` crate before so I don't know where the `AppConfigBuilderError` occurs in the code.

sadsnake42 replied:
Since in this case there are no default values for the fields, an error will occur if someone does not configure `AppConfig`

hipstermojo commented on src/cli.rs:
```diff
@@ -1,13 +1,56 @@
+    UrlFileError(#[from] std::io::Error),
+    #[error("Failed to parse max connection value: {0}")]
+    InvalidMaxConnectionCount(#[from] ParseIntError),
+    #[error("No urls for parse")]
```
Change this to "No urls were provided"

hipstermojo commented on src/cli.rs:
```diff
@@ -1,13 +1,56 @@
+    InvalidOutputPath(String),
+    #[error("Log error: {0}")]
+    LogDirectoryError(String),
+    #[error(transparent)]
```
Change this to
```rs
#[error("Unable to start logger!\n{0}")]
```

hipstermojo commented:
Once everything else is okay also set the PR to merge to `dev` instead of master.

Commit:
Fix from PR#15

- refactor comments
- move `cli::Error` to `errors::ErrorCli`
- removed mixing of order of input urls
- move pure functionality if `init_logger` to clear function

hipstermojo commented on src/errors.rs:
```diff
@@ -124,3 +127,31 @@ impl From<std::str::Utf8Error> for PaperoniError {
+pub enum LogError {
+    #[error(transparent)]
+    FlexiError(#[from] FlexiLoggerError),
+    #[error("Wrong log directory: {0}")]
```
Change this to just:
```rust
#[error("{0}")]
```

cyphersnake replied:
there is no contract that obliges the code that uses this type of error to be about the directory. are you sure you need to reduce the context? maybe I need to correct the error texts in calls of it?

hipstermojo replied:
If you can use a different text for the context then it can stay. I think "Wrong log directory" is a little misleading since one of the possible full errors would be:
`Wrong log directory: Unable to get user directories for logging purposes` which sounds incorrect

cyphersnake replied:
I thought it could be done a little better. Please check

hipstermojo replied:
Looks good now.

Commit:
Refactor `LogError`

hipstermojo commented:
Thanks for making this PR @sadsnake42 :). Given the number of changes made, next time you could have it in individual commits rather than a single commit when you're making significant changes, I prefer it that way. I didn't have contribution guidelines set up so I can't really blame you.

Have a good day!
